File: Services/TaskService.cs
using TodoApp.Models;
using TodoApp.DataAccess.Repositories;
using TodoApp.Services.Interfaces;

namespace TodoApp.Services
{
    public class TaskService : ITaskService
    {
        private readonly IRepository<Tasks> _taskRepository;
        public TaskService(IRepository<Tasks> repository)
        {
            _taskRepository = repository;
        }
        public IEnumerable<Tasks> GetTasks(int pageNumber, int pageSize, int? categoryId, string? searchQuery, int userId)
        {
            IQueryable<Tasks> query = _taskRepository.GetAll().Where(t => t.UserId == userId);

            if (categoryId.HasValue)
            {
                query = query.Where(t => t.CategoryId == categoryId.Value);
            }

            if (!string.IsNullOrWhiteSpace(searchQuery))
            {
                query = query.Where(t => t.Title.ToLower().Contains(searchQuery.ToLower()));
            }

            query = query.Skip((pageNumber - 1) * pageSize).Take(pageSize);
            return query.ToList();
        }
        public Tasks? GetTaskById(int id, int userId)
        {
            return _taskRepository.GetAll().FirstOrDefault(t => t.Id == id && t.UserId == userId);
        }

        public Tasks CreateTask(Tasks task)
        {
            task.CreatedAt = DateTime.Now;
            task.IsCompleted = false;
            _taskRepository.Add(task);
            _taskRepository.SaveChanges();

            return task;
        }

        public void UpdateTask(Tasks task, int userId)
        {
            Tasks? existingTask = GetTaskById(task.Id, userId);
            if (existingTask == null)
            {
                throw new Exception("Task not found!");
            }

            existingTask.Title = task.Title;
            existingTask.Description = task.Description;
            existingTask.DueDate = task.DueDate;
            existingTask.IsCompleted = task.IsCompleted;
            existingTask.CategoryId = task.CategoryId;

            _taskRepository.Update(existingTask);
            _taskRepository.SaveChanges();
        }

        public void DeleteTask(int id, int userId)
        {
            Tasks? task = GetTaskById(id, userId);
            if (task != null)
            {
                _taskRepository.Delete(task);
                _taskRepository.SaveChanges();
            }
        }
    }
}

File: todo-frontend/src/app/services/task.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Task } from '../models/app.models';

@Injectable({
  providedIn: 'root'
})
export class TaskService {
  private apiUrl = 'https://localhost:7015/api/Tasks';

  constructor(private http: HttpClient) {}

  getTasks(pageNumber: number = 1, pageSize: number = 10, categoryId?: number, searchQuery?: string): Observable<Task[]> {
    let params = new HttpParams()
      .set('pageNumber', pageNumber.toString())
      .set('pageSize', pageSize.toString());

    if (categoryId) {
      params = params.set('categoryId', categoryId.toString());
    }
    if (searchQuery) {
      params = params.set('searchQuery', searchQuery);
    }

    return this.http.get<Task[]>(this.apiUrl, { params });
  }

  getTaskById(id: number): Observable<Task> {
    return this.http.get<Task>(`${this.apiUrl}/${id}`);
  }

  createTask(task: Task): Observable<Task> {
    return this.http.post<Task>(this.apiUrl, task);
  }

  updateTask(id: number, task: Task): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${id}`, task);
  }

  deleteTask(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}